File: src/main/java/cz/petrklatik/ratelimitter/ratelimiter/TokenBucket.java
package cz.petrklatik.ratelimitter.ratelimiter;

import java.util.concurrent.atomic.AtomicReference;

/**
 * Token Bucket implementation using CAS (Compare-And-Swap).
 */
public class TokenBucket {

  /**
   * Immutable state snapshot for atomic updates.
   */
  private record State(double tokens, long lastRefillNanos) {
  }

  private final AtomicReference<State> state;
  private final int capacity;
  private final int refillPerSecond;

  public TokenBucket(RateLimitConfig config) {
    this(config, 1.0); // Start with full bucket
  }

  public TokenBucket(RateLimitConfig config, double fillRatio) {
    this.capacity = config.capacity();
    this.refillPerSecond = config.refillPerSecond();
    double initialTokens = capacity * Math.max(0.0, Math.min(1.0, fillRatio));
    this.state = new AtomicReference<>(new State(initialTokens, System.nanoTime()));
  }

  /**
   * Returns current fill ratio (0.0 to 1.0) for preserving state during config updates.
   */
  public double getFillRatio() {
    State current = state.get();
    long now = System.nanoTime();
    double elapsedSeconds = (now - current.lastRefillNanos) / 1_000_000_000.0;
    double refilled = current.tokens + elapsedSeconds * refillPerSecond;
    double currentTokens = Math.min(capacity, refilled);
    return currentTokens / capacity;
  }

  /**
   * Attempts to consume one token from the bucket.
   *
   * @return result with allowed status and rate limit info
   */
  public RateLimitResult tryConsume() {
    while (true) {
      State current = state.get();
      long now = System.nanoTime();

      // Calculate tokens to add based on elapsed time
      double elapsedSeconds = (now - current.lastRefillNanos) / 1_000_000_000.0;
      double refilled = current.tokens + elapsedSeconds * refillPerSecond;
      double newTokens = Math.min(capacity, refilled);

      if (newTokens >= 1.0) {
        // Consume one token
        State next = new State(newTokens - 1.0, now);
        if (state.compareAndSet(current, next)) {
          return RateLimitResult.allowed((int) next.tokens, capacity);
        }
        // Another thread modified state, retry
      } else {
        // Not enough tokens - calculate wait time
        double tokensNeeded = 1.0 - newTokens;
        long waitMillis = (long) Math.ceil((tokensNeeded / refillPerSecond) * 1000);

        // Update timestamp even on rejection
        State next = new State(newTokens, now);
        if (state.compareAndSet(current, next)) {
          return RateLimitResult.rejected(waitMillis, capacity);
        }
        // retry
      }
    }
  }
}

File: src/main/resources/scripts/token_bucket.lua
-- Token Bucket algorithm for Redis
--
-- Runs atomically - no race conditions between instances
--
-- Call: EVALSHA <sha> 1 <key> <capacity> <refill_per_second> <now_millis>
-- Returns: [allowed (0/1), remaining, retry_after_millis]

local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refill_per_second = tonumber(ARGV[2])
local now = tonumber(ARGV[3])

-- Load current state (or initialize with full bucket)
local bucket = redis.call('HMGET', key, 'tokens', 'last_refill')
local tokens = tonumber(bucket[1]) or capacity
local last_refill = tonumber(bucket[2]) or now

-- Calculate refill based on elapsed time
local elapsed = (now - last_refill) / 1000.0
local refilled = math.min(capacity, tokens + elapsed * refill_per_second)

if refilled >= 1 then
    -- Allowed: consume 1 token
    local new_tokens = refilled - 1
    redis.call('HMSET', key, 'tokens', new_tokens, 'last_refill', now)
    redis.call('EXPIRE', key, 3600)
    return {1, math.floor(new_tokens), 0}
else
    -- Rejected: calculate retry time
    local needed = 1 - refilled
    local retry_ms = math.ceil((needed / refill_per_second) * 1000)
    redis.call('HMSET', key, 'tokens', refilled, 'last_refill', now)
    redis.call('EXPIRE', key, 3600)
    return {0, 0, retry_ms}
end
